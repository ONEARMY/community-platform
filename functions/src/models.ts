// tslint:disable no-implicit-dependencies
// Models can be imported from the main package for use here
// NOTE 1 - this requires adjustment main src in package.json
// NOTE 2 - shorthand @OAModels notation defined in tsconfig
import { IDBEndpoint, IDbDoc } from '@OAModels/common.models'

export { IDBEndpoint, IDbDoc }
