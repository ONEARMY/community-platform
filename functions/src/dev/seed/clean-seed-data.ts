import { Request, Response } from "express"
// (note - typings don't currently exist for firebase-tools: https://github.com/firebase/firebase-tools/issues/2378)
import * as firebase_tools from 'firebase-tools'
import { DB_ENDPOINTS } from "../../models"
import { db } from "../../Firebase/firestoreDB"
import { splitArrayToChunks } from "../../Utils/data.utils"
import { firestore, auth } from "firebase-admin"

const USE_SMALL_SAMPLE_SEED = false

/**
 * Script used to generate a cleaner export of seed data for use in development
 * 
 * Given a full dump of site data in the emulator, strips away all user docs and revisions
 * for users that have not posted content to mappins, howtos, research or events
 * (basically anywhere their profile might be linked from)
 */
export async function cleanSeedData(req: Request, res: Response) {
    const dbCollections = await db.listCollections()
    const dbEndpoints = dbCollections.map(c => c.id)
    const expectedEndpoints = Object.values(DB_ENDPOINTS)
    // Delete collections not in use
    for (const endpoint of dbEndpoints) {
        if (!expectedEndpoints.includes(endpoint)) {
            await deleteCollectionCLI(endpoint)
        }
    }

    // setup variables and types for tracking data
    const keptUsers = {}
    const endpointsToCheck = ['mappins', 'howtos', 'research', 'events'] as const
    type ICheckedEndpoint = typeof endpointsToCheck[number]
    const allDocs: { [endpoint in ICheckedEndpoint]: firestore.QuerySnapshot<firestore.DocumentData> } = {} as any
    const returnMessage: { deleted: any, kept: any, created: any } = {
        deleted: {}, kept: {}, created: {}
    }

    // Get list of users with howtos, mappins or events to retain data
    for (const endpoint of endpointsToCheck) {
        const mappedEndpoint = DB_ENDPOINTS[endpoint]
        if (mappedEndpoint) {
            const snapshot = await db.collection(mappedEndpoint).get()
            allDocs[endpoint as any] = snapshot
            snapshot.docs.forEach(doc => {
                switch (endpoint) {
                    case 'mappins':
                        keptUsers[doc.id] = true
                        break;
                    default:
                        const data = doc.data();
                        keptUsers[data._createdBy] = true
                }
            })
        }
    }
    // await WiPReduceSeedSize()



    // Delete non-required users in batches 
    const deletedUsers = await deleteQueryDocs(db.collection(DB_ENDPOINTS.users), '[Users Deleted]', (doc) => !keptUsers[doc.data()._id])
    returnMessage.deleted['users'] = deletedUsers
    // Delete nested revision docs (TODO - should add filter to ensure doc ref is a subcollection of users)
    const deletedRevisions = await deleteQueryDocs(db.collectionGroup('revisions'), '[Revisions Deleted]', (doc) => !keptUsers[doc.data()._id])
    returnMessage.deleted['revisions'] = deletedRevisions
    // Delete nested stats docs
    const deletedStats = await deleteQueryDocs(db.collectionGroup('stats'), '[Stats Deleted]', (doc) => !keptUsers[doc.data()._id])
    returnMessage.deleted['stats'] = deletedStats
    returnMessage.kept.users = keptUsers

    // Add auth users
    const createdUsers = await addSeedDemoAuthUsers()
    returnMessage.created.users = createdUsers

    res.status(200).send(returnMessage)
}

async function addSeedDemoAuthUsers() {
    const authAdmin = auth()
    const authUsers = [{ name: 'demo_user', roles: [] }, { name: 'demo_admin', roles: ['admin'] }]
    const createdUsers = {}
    for (const user of authUsers) {
        try {
            const authResponse = await authAdmin.createUser({ displayName: user.name, password: user.name, uid: user.name, email: `${user.name}@example.com` })
            const userDocRef = db.collection(DB_ENDPOINTS.users).doc(user.name)
            await userDocRef.set({ roles: user.roles })
            createdUsers[user.name] = authResponse
        } catch (error) {
            // user might already exist, should be fine but log for reference
            console.error(error.message)
        }
    }
    return createdUsers
}

/**
 *   WiP - If limiting seed size reduce the overall user list
 *   TODO - this will be too random and also end up with lots more howtos than anything else. Ideally should provide
 */
async function WiPReduceSeedSize(allDocs) {
    // specific list of data
    if (USE_SMALL_SAMPLE_SEED) {
        const sampleSize = 20
        const keptUsers = {}
        const shuffledHowtos = allDocs.howtos.docs.sort(() => 0.5 - Math.random()).slice(0, sampleSize);
        shuffledHowtos.forEach(doc => keptUsers[doc.data()._createdBy] = true)
        // Delete howtos
        const deletedHowtos = await batchDeleteDocs(allDocs.howtos.docs.filter(d => !keptUsers[d.data()._createdBy]), '[Howtos Deleted]')
        // Delete mappins
        const deletedMappins = await batchDeleteDocs(allDocs.mappins.docs.filter(d => !keptUsers[d.id]), '[Mappins Deleted]')
        // Delete events (TBD)
        // Delete research (TBD)
        return { deletedHowtos, deletedMappins, keptUsers }
    }
}

/** Execute a firestore query and delete all retrieved docs, subject to optional filter function */
async function deleteQueryDocs(
    query: firestore.CollectionGroup<firestore.DocumentData> | firestore.CollectionReference<firestore.DocumentData>,
    logPrefix = '[Deleted]',
    filterFn = (doc: firestore.QueryDocumentSnapshot) => false) {
    const queryResults = await query.get()
    const filteredResults = queryResults.docs.filter(doc => filterFn(doc))
    return batchDeleteDocs(filteredResults, logPrefix)
}

async function batchDeleteDocs(docs: firestore.QueryDocumentSnapshot<firestore.DocumentData>[], logPrefix = '[Deleted]',) {
    const refs = docs.map(doc => doc.ref)
    const chunks = splitArrayToChunks(refs, 500)
    let currentChunk = 0
    const totalDocs = refs.length
    for (const chunk of chunks) {
        const batch = db.batch()
        for (const ref of chunk) {
            batch.delete(ref)
        }
        await batch.commit()
        const deleteCount = currentChunk * 500 + chunk.length
        console.log(`${logPrefix} [${deleteCount}] of ${totalDocs}`)
        currentChunk++
        await _waitForNextTick()
    }
}

/** Simple wrapper function to avoid cascading knock-ons for triggered functions etc. */
function _waitForNextTick(): Promise<void> {
    return new Promise((resolve) => process.nextTick(() => resolve()))
}

/**
 * Use firebase tools to fully delete collection and subcollection from command-line tools
 * https://firebase.google.com/docs/firestore/solutions/delete-collections
 */
async function deleteCollectionCLI(endpoint: string) {
    // Note - whilst we are only operating on the emulator user will probably
    // still need to be logged into firebase to call
    // https://github.com/firebase/firebase-tools/issues/1940
    await firebase_tools.firestore.delete(endpoint, {
        recursive: true,
        yes: true,
    });
}


/**
 * Iterative method to delete all documents in a collection, by reading all documents in batches and deleting
 * Copied from https://github.com/firebase/snippets-node/blob/e5f6214059bdbc63f94ba6600f7f84e96325548d/firestore/main/index.js#L889-L921
 *
 * Note 1 - this is less efficient than the cli method, but also available outside of node environment
 * Note 2 - This fails to delete subcollections, and manual methods above added instead
 */
async function deleteCollectionIteratively(collectionPath: string, batchSize = 500) {
    const collectionRef = db.collection(collectionPath);
    const query = collectionRef.orderBy('__name__').limit(batchSize);
    return new Promise((resolve, reject) => {
        deleteQueryBatch(query, resolve).catch(reject);
    });
}

async function deleteQueryBatch(query, resolve) {
    const snapshot = await query.get();

    const batchSize = snapshot.size;
    if (batchSize === 0) {
        // When there are no documents left, we are done
        resolve();
        return;
    }

    // Delete documents in a batch
    const batch = db.batch();
    snapshot.docs.forEach((doc) => {
        batch.delete(doc.ref);
    });
    await batch.commit();

    // Recurse on the next process tick, to avoid
    // exploding the stack.
    process.nextTick(async () => {
        await deleteQueryBatch(query, resolve);
    });
}
