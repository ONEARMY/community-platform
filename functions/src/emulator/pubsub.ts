import { PubSub } from '@google-cloud/pubsub'

interface ITriggerOptions {
  trigger: 'topic' | 'schedule'
  name: string
  payload: any
}
const defaultOptions: ITriggerOptions = {
  trigger: 'topic',
  name: 'test-topic',
  payload: {},
}

/**
 * Trigger a pubsub message
 * @param options.trigger - specify as 'schedule' if trying to call a scheduled function.
 * Default 'topic'
 * @param options.name - name of the topic to be scheduled. This will either be the name of the
 * scheduled function, or the name of the pubsub topic to trigger. Default "test-topic"
 * @param options.payload - optional payload passed to function. Default {}
 * @returns confirmation messageID of triggered pubsub
 */
export const triggerPubsub = async (options: ITriggerOptions) => {
  let { name, payload, trigger } = { ...defaultOptions, ...options }
  const pubsub = new PubSub()
  if (trigger === 'schedule') {
    // scheduled functions can be triggered by adding prefix to topic name
    // https://github.com/firebase/firebase-tools/pull/2011/files
    name = `firebase-schedule-${name}`
  }

  const [topics] = await pubsub.getTopics()
  if (!topics.find((t) => t.name.includes(name))) {
    await pubsub.createTopic(name)
  }

  const messageID = await pubsub.topic(name).publishMessage({ json: payload })
  return messageID
}
