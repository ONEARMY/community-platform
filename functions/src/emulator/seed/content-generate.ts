import { IMockAuthUser, MOCK_AUTH_USERS } from 'oa-shared/mocks/auth'
import type { IHowtoDB, IUserDB } from '../../models'
import { setDoc, updateDoc } from '../../Firebase/firestoreDB'

/**
 * Populate additional mock howtos alongside production data for ease of testing
 * TODO - should create from factories used in unit mocks once available in shared
 */
export async function seedContentGenerate() {
  // create mock howtos just for demo_beta_tester and demo_admin users
  for (const user of Object.values(MOCK_AUTH_USERS).slice(1, 3)) {
    await setMockHowto(user)
    await setMockNotifications(user)
  }
  return
}

async function setMockHowto(user: IMockAuthUser) {
  const { uid } = user
  const _id = `00_${uid}_howto`
  const loginInfo = `username : ${uid}@example.com\npassword : ${uid}`
  const howto: IHowtoDB = {
    _id,
    _created: new Date().toISOString(),
    _modified: new Date().toISOString(),
    _deleted: false,
    _createdBy: uid,
    cover_image: {
      downloadUrl: `https://platform.onearmy.earth/images/One-Army-Star-Logo.svg`,
    } as any,
    description: `You can edit this howto by logging in as:\n\n${loginInfo}`,
    difficulty_level: 'Easy',
    files: [],
    mentions: [],
    slug: _id,
    steps: [],
    time: '',
    title: 'Mock Howto',
    moderation: 'accepted',
    comments: [
      {
        _created: new Date().toISOString(),
        _creatorId: 'demo_user',
        _id: 'mock_comment',
        creatorName: 'demo_user',
        text: 'Mock generated comment',
      },
    ],
  }
  await setDoc('howtos', _id, howto)
}

async function setMockNotifications(user: IMockAuthUser) {
  const update: Partial<IUserDB> = {
    notifications: [
      {
        type: 'new_comment',
        notified: false,
        read: false,
        _created: new Date().toISOString(),
        _id: 'mock_notification',
        triggeredBy: { displayName: 'Demo User', userId: 'demo_user' },
      },
    ],
    notification_settings: {
      emailFrequency: 'weekly' as any, // should set from Enum but want to avoid import
      enabled: {
        new_comment: true,
        howto_useful: true,
        howto_mention: true,
        new_comment_research: true,
        research_useful: true,
        research_mention: true,
      },
    },
  }
  await updateDoc('users', user.uid, update)
}
