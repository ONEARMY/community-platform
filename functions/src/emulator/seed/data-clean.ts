// (note - typings don't currently exist for firebase-tools: https://github.com/firebase/firebase-tools/issues/2378)
import * as firebase_tools from 'firebase-tools'
import { DB_ENDPOINTS } from '../../models'
import { db } from '../../Firebase/firestoreDB'
import { splitArrayToChunks } from '../../Utils/data.utils'
import type { firestore } from 'firebase-admin'
import axios from 'axios'

const USE_SMALL_SAMPLE_SEED = false

/**
 * Script used to generate a cleaner export of seed data for use in development
 *
 * Given a full dump of site data in the emulator, strips away all user docs and revisions
 * for users that have not posted content to mappins, howtos, research or events
 * (basically anywhere their profile might be linked from)
 */
export async function seedDataClean() {
  const dbCollections = await db.listCollections()
  const dbEndpoints = dbCollections.map((c) => c.id)
  console.log('db endpoints', dbEndpoints)
  const expectedEndpoints = Object.values(DB_ENDPOINTS)
  const returnMessage = {
    endpoints: { deleted: {} as any, kept: {} as any },
  }

  // Delete collections not in use
  for (const endpoint of dbEndpoints) {
    if (!expectedEndpoints.includes(endpoint)) {
      await deleteCollectionAPI(endpoint)
      returnMessage.endpoints.deleted[endpoint] = 'All'
    } else {
      returnMessage.endpoints.kept[endpoint] = 'All'
    }
  }

  // setup variables and types for tracking data
  const keptUsers = {}
  const endpointsToCheck = ['mappins', 'howtos', 'research', 'events'] as const
  type ICheckedEndpoint = typeof endpointsToCheck[number]
  const allDocs: {
    [endpoint in ICheckedEndpoint]: firestore.QuerySnapshot<firestore.DocumentData>
  } = {} as any

  // Get list of users with howtos, mappins or events to retain data
  for (const endpoint of endpointsToCheck) {
    const mappedEndpoint = DB_ENDPOINTS[endpoint]
    if (mappedEndpoint) {
      const snapshot = await db.collection(mappedEndpoint).get()
      allDocs[endpoint as any] = snapshot
      snapshot.docs.forEach((doc) => {
        switch (endpoint) {
          case 'mappins':
            keptUsers[doc.id] = true
            break
          default:
            const data = doc.data()
            keptUsers[data._createdBy] = true
        }
      })
    }
  }
  // await WiPReduceSeedSize()

  // Delete non-required users in batches
  const deletedUsers = await deleteQueryDocs(
    db.collection(DB_ENDPOINTS.users),
    '[Users Deleted]',
    (doc) => !keptUsers[doc.data()._id],
  )
  returnMessage.endpoints.deleted.users = deletedUsers.length
  // Delete nested revision docs (TODO - should add filter to ensure doc ref is a subcollection of users)
  const deletedRevisions = await deleteQueryDocs(
    db.collectionGroup('revisions'),
    '[Revisions Deleted]',
    (doc) => !keptUsers[doc.data()._id],
  )
  returnMessage.endpoints.deleted['revisions'] = deletedRevisions.length
  // Delete nested stats docs
  const deletedStats = await deleteQueryDocs(
    db.collectionGroup('stats'),
    '[Stats Deleted]',
    (doc) => !keptUsers[doc.data()._id],
  )
  returnMessage.endpoints.deleted['stats'] = deletedStats.length
  returnMessage.endpoints.kept.users = Object.keys(keptUsers).length

  return returnMessage
}

/**
 *   WiP - If limiting seed size reduce the overall user list
 *   TODO - this will be too random and also end up with lots more howtos than anything else. Ideally should provide
 */
async function WiPReduceSeedSize(allDocs) {
  // specific list of data
  if (USE_SMALL_SAMPLE_SEED) {
    const sampleSize = 20
    const keptUsers = {}
    const shuffledHowtos = allDocs.howtos.docs
      .sort(() => 0.5 - Math.random())
      .slice(0, sampleSize)
    shuffledHowtos.forEach((doc) => (keptUsers[doc.data()._createdBy] = true))
    // Delete howtos
    const deletedHowtos = await batchDeleteDocs(
      allDocs.howtos.docs.filter((d) => !keptUsers[d.data()._createdBy]),
      '[Howtos Deleted]',
    )
    // Delete mappins
    const deletedMappins = await batchDeleteDocs(
      allDocs.mappins.docs.filter((d) => !keptUsers[d.id]),
      '[Mappins Deleted]',
    )
    // Delete events (TBD)
    // Delete research (TBD)
    return { deletedHowtos, deletedMappins, keptUsers }
  }
}

/** Execute a firestore query and delete all retrieved docs, subject to optional filter function */
async function deleteQueryDocs(
  query:
    | firestore.CollectionGroup<firestore.DocumentData>
    | firestore.CollectionReference<firestore.DocumentData>,
  logPrefix = '[Deleted]',
  filterFn = (doc: firestore.QueryDocumentSnapshot) => false,
) {
  const queryResults = await query.get()
  const filteredResults = queryResults.docs.filter((doc) => filterFn(doc))
  await batchDeleteDocs(filteredResults, logPrefix)
  return filteredResults
}

async function batchDeleteDocs(
  docs: firestore.QueryDocumentSnapshot<firestore.DocumentData>[],
  logPrefix = '[Deleted]',
) {
  const refs = docs.map((doc) => doc.ref)
  const chunks = splitArrayToChunks(refs, 500)
  let currentChunk = 0
  const totalDocs = refs.length
  for (const chunk of chunks) {
    const batch = db.batch()
    for (const ref of chunk) {
      batch.delete(ref)
    }
    await batch.commit()
    const deleteCount = currentChunk * 500 + chunk.length
    console.log(`${logPrefix} [${deleteCount}] of ${totalDocs}`)
    currentChunk++
    await _waitForNextTick()
  }
}

/** Simple wrapper function to avoid cascading knock-ons for triggered functions etc. */
function _waitForNextTick(): Promise<void> {
  return new Promise((resolve) => process.nextTick(() => resolve()))
}

/**
 * Use rest api to delete documents being served on the emulator
 * @param endpoint
 * @returns
 */
async function deleteCollectionAPI(endpoint: string) {
  const apiHost =
    'http://0.0.0.0:4003/emulator/v1/projects/community-platform-emulated' // http://[::1] for non-docker env
  return axios.delete(`${apiHost}/databases/(default)/documents/${endpoint}`)
}

/**
 * Use firebase tools to fully delete collection and subcollection from command-line tools
 * https://firebase.google.com/docs/firestore/solutions/delete-collections
 *
 * NOTE CC 2021-09-29 - Seems to have stopped working with current emulators (not sure why)
 * reverting to api method instead. Might be fixed with updated host endpoint used above and in firebase.json
 */
async function deleteCollectionCLI(endpoint: string) {
  // Note - whilst we are only operating on the emulator user will probably
  // still need to be logged into firebase to call
  // https://github.com/firebase/firebase-tools/issues/1940
  await firebase_tools.firestore
    .delete(endpoint, {
      recursive: true,
      yes: true,
    })
    .catch((err) => {
      console.error(err)
      process.exit(1)
    })
}

/**
 * Iterative method to delete all documents in a collection, by reading all documents in batches and deleting
 * Copied from https://github.com/firebase/snippets-node/blob/e5f6214059bdbc63f94ba6600f7f84e96325548d/firestore/main/index.js#L889-L921
 *
 * Note 1 - this is less efficient than the cli method, but also available outside of node environment
 * Note 2 - This fails to delete subcollections, and manual methods above added instead
 */
async function deleteCollectionIteratively(
  collectionPath: string,
  batchSize = 500,
) {
  const collectionRef = db.collection(collectionPath)
  const query = collectionRef.orderBy('__name__').limit(batchSize)
  return new Promise((resolve, reject) => {
    deleteQueryBatch(query, resolve).catch(reject)
  })
}

async function deleteQueryBatch(query, resolve) {
  const snapshot = await query.get()

  const batchSize = snapshot.size
  if (batchSize === 0) {
    // When there are no documents left, we are done
    resolve()
    return
  }

  // Delete documents in a batch
  const batch = db.batch()
  snapshot.docs.forEach((doc) => {
    batch.delete(doc.ref)
  })
  await batch.commit()

  // Recurse on the next process tick, to avoid
  // exploding the stack.
  process.nextTick(async () => {
    await deleteQueryBatch(query, resolve)
  })
}
