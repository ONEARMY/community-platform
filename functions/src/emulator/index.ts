import * as functions from 'firebase-functions'
import cors from 'cors'
import express from 'express'
import { seedDataClean } from './seed/data-clean'
import { seedUsersCreate } from './seed/users-create'
import { triggerPubsub } from './pubsub'

console.log('emulator api ready')

const app = express()
app.use(cors({ origin: true }))

/**
 * Expose an api endpoint available just for use with emulators
 * NOTE - if changing endpoints check emulators workspace in case in use (e.g. called via Dockerfile)
 */
app.use(function (req, res, next) {
  if (process.env.FUNCTIONS_EMULATOR === 'true') next()
  else
    res
      .status(403)
      .send(
        `Emulator api methods can only be accessed in emulator environment [${process.env.FUNCTIONS_EMULATOR}]`,
      )
})

app.get('/', (req, res) => res.status(200).send('Emulator Api Working'))

app.get('/seed-users-create', (req, res) =>
  seedUsersCreate().then((users) =>
    res
      .status(200)
      .send(`Created users: ${users.map((u) => u.uid).join(', ')}`),
  ),
)

app.get('/seed-clean', (req, res) =>
  seedDataClean().then((result) => {
    res.status(200).send(result)
  }),
)

/**
 * As there is currently no UI to publish pubsub messages for testing, create
 * a http wrapper that can be used to call via rest
 * https://github.com/firebase/firebase-tools/issues/2034
 * https://blog.minimacode.com/publish-message-to-pubsub-emulator/
 *
 * This allows pubsub topics to be tested when running in emulators as an api endpoint
 * It includes additional prefix (/topic or /schedule) depending on trigger
 * @example
 * `http://localhost:4002/community-platform-emulated/us-central1/emulator/pubsub/schedule/dailyTasks`
 * `http://localhost:4002/community-platform-emulated/us-central1/emulator/pubsub/topic/test-topic`
 */
app.all('/pubsub/:trigger/:name', async (req, res) => {
  if (!process.env.PUBSUB_EMULATOR_HOST) {
    functions.logger.error(
      'This function should only run locally in an emulator.',
    )
    res.status(400).end()
  }
  const trigger = req.params.trigger as 'topic' | 'schedule'
  const name = req.params.name
  const payload = req.body
  const messageId = await triggerPubsub({ trigger, name, payload })
  const result = 'Pubsub Published'
  res.status(201).send({ result, trigger, name, messageId, payload })
})

// Increase default timeout to max allowed for longer-running operations
export = functions.runWith({ timeoutSeconds: 540 }).https.onRequest(app as any)
