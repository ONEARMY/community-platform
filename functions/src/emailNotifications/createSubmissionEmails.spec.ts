import { FirebaseEmulatedTest } from '../test/Firebase/emulator'
import { DB_ENDPOINTS, IMapPin, IUserDB } from '../models'
import {
  HOW_TO_SUBMISSION_SUBJECT,
  MAP_PIN_SUBMISSION_SUBJECT,
} from './templates'
import { setMockHowto } from '../emulator/seed/content-generate'
import {
  createHowtoSubmissionEmail,
  createMapPinSubmissionEmail,
} from './createSubmissionEmails'

jest.mock('../Firebase/auth', () => ({
  firebaseAuth: {
    getUser: () => ({
      email: 'test@test.com',
    }),
  },
}))

jest.mock('../config/config', () => ({
  CONFIG: {
    deployment: {
      site_url: 'https://community.preciousplastic.com',
    },
  },
}))

const userFactory = (_id: string, user: Partial<IUserDB> = {}): IUserDB =>
  ({
    _id,
    _authID: _id,
    ...user,
  } as IUserDB)

describe('Create howto submission emails', () => {
  const db = FirebaseEmulatedTest.admin.firestore()

  beforeAll(async () => {
    await FirebaseEmulatedTest.clearFirestoreDB()
    await FirebaseEmulatedTest.seedFirestoreDB('emails')

    await FirebaseEmulatedTest.seedFirestoreDB('users', [
      userFactory('user_1', {
        displayName: 'User 1',
        userName: 'user_1',
        userRoles: ['beta-tester'],
      }),
      userFactory('user_2', {
        displayName: 'User 2',
        userName: 'user_2',
      }),
    ])

    await FirebaseEmulatedTest.seedFirestoreDB('howtos')
  })

  afterAll(async () => {
    await FirebaseEmulatedTest.clearFirestoreDB()
  })

  it('Creates an email for a submitted howto', async () => {
    const howto = await setMockHowto({ uid: 'user_1' }, 'awaiting-moderation')
    await createHowtoSubmissionEmail(howto)

    // Only one submitted howto email should have been created
    const countSnapshot = await db.collection(DB_ENDPOINTS.emails).count().get()
    expect(countSnapshot.data().count).toEqual(1)

    const querySnapshot = await db.collection(DB_ENDPOINTS.emails).get()
    querySnapshot.forEach((doc) => {
      const {
        message: { html, subject },
        to,
      } = doc.data()
      expect(html).toMatchSnapshot()
      expect(subject).toBe(HOW_TO_SUBMISSION_SUBJECT)
      // Check that the email contains the correct user name
      expect(html).toContain('Hey User 1')
      // Check that the email contains the correct howto title
      expect(html).toContain(
        `Huzzah! Your How-To Mock Howto has been submitted.`,
      )
      // Check that the email contains the correct PP signoff
      expect(html).toContain('Charlie your Precious Plastic Community Manager')
      expect(to).toBe('test@test.com')
    })
  })

  it('Does not create email for draft how tos', async () => {
    const howto = await setMockHowto({ uid: 'user_2' }, 'draft')
    await createHowtoSubmissionEmail(howto)

    // No new emails should have been created
    const countSnapshot = await db.collection(DB_ENDPOINTS.emails).count().get()
    expect(countSnapshot.data().count).toEqual(1)
  })

  // Remove this test once released to all users.
  it('Does not create email for people who are not beta testers', async () => {
    const howto = await setMockHowto({ uid: 'user_2' })
    await createHowtoSubmissionEmail(howto)

    // No new emails should have been created
    const countSnapshot = await db.collection(DB_ENDPOINTS.emails).count().get()
    expect(countSnapshot.data().count).toEqual(1)
  })
})

describe('Create map pin submission emails', () => {
  const db = FirebaseEmulatedTest.admin.firestore()

  beforeAll(async () => {
    await FirebaseEmulatedTest.clearFirestoreDB()
    await FirebaseEmulatedTest.seedFirestoreDB('emails')

    await FirebaseEmulatedTest.seedFirestoreDB('users', [
      userFactory('user_1', {
        displayName: 'User 1',
        userName: 'user_1',
        userRoles: ['beta-tester'],
      }),
      userFactory('user_2', {
        displayName: 'User 2',
        userName: 'user_2',
      }),
    ])

    await FirebaseEmulatedTest.seedFirestoreDB('mappins')
  })

  afterAll(async () => {
    await FirebaseEmulatedTest.clearFirestoreDB()
  })

  it('Creates an email for a submitted map pin', async () => {
    const mapPin = {
      _id: 'user_1',
      moderation: 'awaiting-moderation',
    }
    await createMapPinSubmissionEmail(mapPin as IMapPin)

    // Only one submitted map pin email should have been created
    const countSnapshot = await db.collection(DB_ENDPOINTS.emails).count().get()
    expect(countSnapshot.data().count).toEqual(1)

    const querySnapshot = await db.collection(DB_ENDPOINTS.emails).get()
    querySnapshot.forEach((doc) => {
      const {
        message: { html, subject },
        to,
      } = doc.data()
      expect(html).toMatchSnapshot()
      expect(subject).toBe(MAP_PIN_SUBMISSION_SUBJECT)
      // Check that the email contains the correct user name
      expect(html).toContain('Hey User 1')
      // Check that the email contains the correct title
      expect(html).toContain('Your map pin has been submitted.')
      // Check that the email contains the correct PP signoff
      expect(html).toContain('Charlie your Precious Plastic Community Manager')
      expect(to).toBe('test@test.com')
    })
  })

  // Remove this test once released to all users.
  it('Does not creates email for people who are not beta testers', async () => {
    const howto = await setMockHowto({ uid: 'user_2' })
    await createHowtoSubmissionEmail(howto)

    // No new emails should have been created
    const countSnapshot = await db.collection(DB_ENDPOINTS.emails).count().get()
    expect(countSnapshot.data().count).toEqual(1)
  })
})
