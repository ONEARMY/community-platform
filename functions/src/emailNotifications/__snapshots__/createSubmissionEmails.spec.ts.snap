// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create howto submission emails Creates an email for a submitted howto 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html charset=UTF-8" />
    <style>
      #body {
        font-family: Arial, Helvetica, sans-serif;
        font-size: 14px;
        color: #000000;
      }
      #body a {
        color: inherit;
      }
      .border {
        border: 3px solid black;
        border-radius: 10px;
        padding: 4% 0;
        margin: 4% 0;
        width: 550px;
      }
      table {
        width: 100%;
        border: 0;
      }
      .settings-table-container,
      .project-image-table-container {
        margin-bottom: 8%;
      }

      .greeting-container,
      .main-container {
        margin-bottom: 8%;
      }
      .greeting-container,
      .main-container,
      .closing-containter {
        margin-left: 8%;
        margin-right: 8%;
      }
      @media only screen and (max-width: 550px) {
        .table-container {
          width: 100%;
        }
      }
    </style>
  </head>
  <body id="body">
    <table class="email-table-container">
      <tr>
        <td align="center">
          <div class="border">
            <table class="project-image-table-container">
              <tr>
                <td align="center">
                  <img
                    width="144"
                    alt=""
                    src="http://cdn.mcauto-images-production.sendgrid.net/cb5a685b085b7e7c/f48e2455-6dbb-4102-bde9-2ee07d3008c4/859x860.png"
                  />
                </td>
              </tr>
            </table>

            <div align="left" class="greeting-container">
              <p>Hey User 1</p>
              <p>Huzzah! Your How-To Mock Howto has been submitted.</p>
            </div>
            <div align="left" class="main-container">
              <p>
                Our team is currently reviewing it and will get back to you
                within a few business days.
              </p>
              <p>
                Thank you for taking the time, the community will benefit from
                this knowledge and more plastic will be recycled thanks to YOU.
              </p>
            </div>
            <div align="left" class="closing-containter">
              <p>Cheers,</p>
              <p>Charlie your Precious Plastic Community Manager</p>
            </div>
          </div>
          <table class="settings-table-container">
            <tr>
              <td align="center">
                <div class="notifications">
                  Manage your notifications
                  <a href="https://community.preciousplastic.com/settings">
                    here</a
                  >
                </div>
              </td>
            </tr>
          </table>
        </td>
      </tr>
    </table>
  </body>
</html>
"
`;

exports[`Create map pin submission emails Creates an email for a submitted map pin 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html charset=UTF-8" />
    <style>
      #body {
        font-family: Arial, Helvetica, sans-serif;
        font-size: 14px;
        color: #000000;
      }
      #body a {
        color: inherit;
      }
      .border {
        border: 3px solid black;
        border-radius: 10px;
        padding: 4% 0;
        margin: 4% 0;
        width: 550px;
      }
      table {
        width: 100%;
        border: 0;
      }
      .settings-table-container,
      .project-image-table-container {
        margin-bottom: 8%;
      }

      .greeting-container,
      .main-container {
        margin-bottom: 8%;
      }
      .greeting-container,
      .main-container,
      .closing-containter {
        margin-left: 8%;
        margin-right: 8%;
      }
      @media only screen and (max-width: 550px) {
        .table-container {
          width: 100%;
        }
      }
    </style>
  </head>
  <body id="body">
    <table class="email-table-container">
      <tr>
        <td align="center">
          <div class="border">
            <table class="project-image-table-container">
              <tr>
                <td align="center">
                  <img
                    width="144"
                    alt=""
                    src="http://cdn.mcauto-images-production.sendgrid.net/cb5a685b085b7e7c/f48e2455-6dbb-4102-bde9-2ee07d3008c4/859x860.png"
                  />
                </td>
              </tr>
            </table>

            <div align="left" class="greeting-container">
              <p>Hey User 1</p>
              <p>Your map pin has been submitted.</p>
            </div>
            <div align="left" class="main-container">
              <p>
                Our team is currently reviewing it and will get back to you
                within a few business days.
              </p>
            </div>
            <div align="left" class="closing-containter">
              <p>Cheers,</p>
              <p>Charlie your Precious Plastic Community Manager</p>
            </div>
          </div>
          <table class="settings-table-container">
            <tr>
              <td align="center">
                <div class="notifications">
                  Manage your notifications
                  <a href="https://community.preciousplastic.com/settings">
                    here</a
                  >
                </div>
              </td>
            </tr>
          </table>
        </td>
      </tr>
    </table>
  </body>
</html>
"
`;
